x/gateway/keeper/unbond_gateways.go

```diff
@@ -0,0 +1,93 @@
+package keeper
+
+import (
+	"context"
+	"fmt"
+
+	cosmostypes "github.com/cosmos/cosmos-sdk/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	gatewaytypes "github.com/pokt-network/poktroll/x/gateway/types"
+	sharedtypes "github.com/pokt-network/poktroll/x/shared/types"
+)
+
+// EndBlockerUnbondGateways unbonds gateways whose unbonding period has elapsed.
+func (k Keeper) EndBlockerUnbondGateways(ctx context.Context) error {
+	logger := k.Logger().With("method", "EndBlockerUnbondGateways")
+
+	sdkCtx := cosmostypes.UnwrapSDKContext(ctx)
+	sharedParams := k.sharedKeeper.GetParams(sdkCtx)
+	currentHeight := sdkCtx.BlockHeight()
+
+	// Only process unbonding gateways at the end of the session.
+	if sharedtypes.IsSessionEndHeight(&sharedParams, currentHeight) {
+		return nil
+	}
+
+	// Iterate over all gateways and unbond the ones that have finished the unbonding period.
+	// TODO_IMPROVE: Use an index to iterate over the gateways that have initiated
+	// the unbonding action instead of iterating over all of them.
+	for _, gateway := range k.GetAllGateways(ctx) {
+		// Ignore gateways that have not initiated the unbonding action.
+		if !gateway.IsUnbonding() {
+			continue
+		}
+
+		unbondingEndHeight := gatewaytypes.GetGatewayUnbondingHeight(&sharedParams, &gateway)
+
+		// If the unbonding height is ahead of the current height, the application
+		// stays in the unbonding state.
+		if unbondingEndHeight > currentHeight {
+			continue
+		}
+
+		if err := k.UnbondGateway(ctx, &gateway); err != nil {
+			return err
+		}
+
```

Comment:
Don't we need to also update applications here?

![Screenshot 2025-02-12 at 5 50 17 PM](https://github.com/user-attachments/assets/022bd8c7-8f59-4038-8633-94ce6e339640)

Reply:
Please #PUC